src/validatefs/validatefs.c

```diff
@@ -0,0 +1,308 @@
+                        return -EINVAL;
+
+                default:
+                        assert_not_reached();
+                }
+
+        if (optind + 1 != argc)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                       "%s excepts exactly one argument (the mount point).",
+                                       program_invocation_short_name);
+
+        arg_target = strdup(argv[optind]);
+        if (!arg_target)
+                return log_oom();
+
+        if (arg_root && !path_startswith(arg_target, arg_root))
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Specified path '%s' does not start with specified root '%s', refusing.", arg_target, arg_root);
+
+        return 1;
+}
+
+typedef struct ValidateFields {
+        sd_id128_t gpt_type_uuid;
+        char *gpt_label;
+        char *mount_point;
+} ValidateFields;
+
+static void validate_fields_done(ValidateFields *f) {
+        assert(f);
+
+        free(f->gpt_label);
+        free(f->mount_point);
+}
+
+static int validate_fields_read(int fd, ValidateFields *ret) {
+        _cleanup_(validate_fields_done) ValidateFields f = {};
+        int r;
+
+        assert(fd >= 0);
+        assert(ret);
+
+        _cleanup_free_ char *t = NULL;
+        r = getxattr_at_malloc(fd, /* path= */ NULL, "user.validatefs.gpt_type_uuid", AT_EMPTY_PATH, &t);
+        if (r < 0) {
+                if (r != -ENODATA && !ERRNO_IS_NOT_SUPPORTED(r))
+                        return log_error_errno(r, "Failed to read 'user.validatefs.gpt_type_uuid' xattr: %m");
+        } else {
+                r = sd_id128_from_string(t, &f.gpt_type_uuid);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to parse 'user.validatefs.gpt_type_uuid' xattr: %s", t);
+        }
+
+        r = getxattr_at_malloc(fd, /* path= */ NULL, "user.validatefs.gpt_label", AT_EMPTY_PATH, &f.gpt_label);
+        if (r < 0) {
+                if (r != -ENODATA && !ERRNO_IS_NOT_SUPPORTED(r))
+                        return log_error_errno(r, "Failed to read 'user.validatefs.gpt_label' xattr: %m");
+        } else if (!utf8_is_valid(f.gpt_label) || string_has_cc(f.gpt_label, /* ok= */ NULL))
+                return log_error_errno(
+                                SYNTHETIC_ERRNO(EINVAL),
+                                "Extended attribute 'user.validatefs.gpt_label' contains invalid characters, refusing.");
+
+        r = getxattr_at_malloc(fd, /* path= */ NULL, "user.validatefs.mount_point", AT_EMPTY_PATH, &f.mount_point);
+        if (r < 0) {
+                if (r != -ENODATA && !ERRNO_IS_NOT_SUPPORTED(r))
+                        return log_error_errno(r, "Failed to read 'user.validatefs.mount_point' xattr: %m");
+        } else if (!utf8_is_valid(f.mount_point) ||
+                   string_has_cc(f.mount_point, /* ok= */ NULL) ||
+                   !path_is_valid(f.mount_point) ||
+                   !path_is_absolute(f.mount_point))
```

Comment:
The check is simpler than `path_is_valid()` (or friends). Please swap the order.